### DISCUSSION TITLE
fillScreen only partial with ST7789VI and ESP32-S3

### DISCUSSION THREAD

**Discussion by gunstr** [Bodmer/TFT_eSPI#3587] [Compatible displays and setup files]
I have just started to test this library and have seen a problem similar to #2933 but with different hardware. I use the Newhaven display [NHD-2.4-240320CF-BSXV-FT](https://newhavendisplay.com/2-4-inch-premium-spi-resistive-tft-display/) that comes with an ST7789VI driver. As a controller I use an ESP32-S3-DEVKITC-1 board. The tests I have done so far are successful except that the `fillScreen` method only fills the screen partially like this:

![20241218_144628](https://github.com/user-attachments/assets/121aa1ba-6334-4ff4-94ea-d1ab11c34662)

I'm using PlatformIO with build flags as below:

```
build_flags =
  -Os
  -DCORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_DEBUG
  -DUSER_SETUP_LOADED=1
  
  ; Define the TFT driver, pins etc. here:
  -DST7789_2_DRIVER=1
  -DTFT_INVERSION_OFF=1
  -DTFT_WIDTH=320
  -DTFT_HEIGHT=240
  -DTFT_MISO=13
  -DTFT_MOSI=11
  -DTFT_SCLK=12
  -DTFT_CS=10
  -DTFT_DC=9
  -DTFT_RST=-1
  ;-DTFT_BL=21
  ;-DTOUCH_CS=22
  -DLOAD_GLCD=1
  -DLOAD_FONT2=1
  -DLOAD_FONT4=1
  -DLOAD_FONT6=1
  -DLOAD_FONT7=1
  -DLOAD_FONT8=1
  -DLOAD_GFXFF=1
  -DSMOOTH_FONT=1
  -DSPI_FREQUENCY=40000000
  -DSPI_READ_FREQUENCY=6000000
```

To narrow down the issue the picture above is captured running the little code snippet from #2933:

```
void setup()
{
  tft.begin();
  tft.setRotation(1);
  tft.fillScreen(TFT_BLACK);
  tft.setRotation(1);
  tft.fillScreen(TFT_RED);
  tft.setCursor(10, 10);
  tft.setTextColor(TFT_WHITE, TFT_BLACK);
  tft.println("Hello World!");
}

void loop()
{
}

```
In contrast to #2933 I see the same problem with any rotation (0-3). The only difference is that with rotation = 0 and 1 the rightmost part of the screen is not filled while with rotation = 2 and 3 it's the leftmost part.

What I understood from the disucssion the root cause might be related to clock speed / timing why I have tested to reduce the clock frequency from 40MHz to 20MHz and also 10MHz but with the same result.

**Comment by Blitz54** [comment]
Try adding this to your code, replace the usual "TFT_eSPI tft = TFT_eSPI;". Had the "same" problem last week, I used ILI9341_DRIVER, and Setup42_ILI9341_ESP32.h in User_Setups. I also had to change the colour order of my screen, but not sure if you'll need that yet. I set the Rotation to 3, but again yours may be different.
`TFT_eSPI tft = TFT_eSPI(320, 240);`

EDIT: You probably don't need the 9341 setup like I did then, didn't notice you have the actual driver listed on the store listing. My screen was a cheapo that didn't have it listed, so it was guess work.

**Comment by gunstr** [comment]
Thanks a lot for the suggestion @Blitz54.

With the hw I use the values actually have to be swapped:
`TFT_eSPI tft = TFT_eSPI(320, 240);`

And yes - I also had to invert the colors with `DTFT_INVERSION_OFF=1`

I have tested with the ArchFill and AllFreeFonts examples and so far it then works like a charm.

What puzzles me a bit is that in the settings I have to configure  `DTFT_WIDTH=320` and `DTFT_HEIGHT=240`, but if I check the constructor for the TFT_eSPI class the parameters are actually the other way around:

https://github.com/Bodmer/TFT_eSPI/blob/5793878d24161c1ed23ccb136f8564f332506d53/TFT_eSPI.h#L426-L432

The root cause for this I have not managed to figure out, but anyway -  what I finally did is to use the settings parameters in the call like this:

`TFT_eSPI tft = TFT_eSPI(TFT_HEIGHT, TFT_WIDTH);`
